apps/system/js/nfc_manager.js

```diff
@@ -200,134 +186,67 @@
-      var state;
-      switch (evt.type) {
-        case 'lockscreen-appopened': // Fall through
-        case 'lockscreen-appclosed':
-        case 'screenchange':
-          if (!this.isActive()) {
-            return;
-          }
-          state = (ScreenManager.screenEnabled && !Service.locked) ?
-                    this.NFC_HW_STATE.ENABLE_DISCOVERY :
-                    this.NFC_HW_STATE.DISABLE_DISCOVERY;
-          if (state === this._hwState) {
-            return;
-          }
-          this._changeHardwareState(state);
-          break;
-        case 'shrinking-sent':
-          window.removeEventListener('shrinking-sent', this);
-          // Notify lower layers that User has acknowledged to send NDEF msg
-          this._dispatchP2PUserResponse();
-
-          // Stop the P2P UI
-          window.dispatchEvent(new CustomEvent('shrinking-stop'));
-          break;
+    _doNfcStateTransition: function(evt) {
+      var evtIdx = NfcHwEvents.indexOf(evt);
+      var state = NfcHwStateTable[this._hwState][evtIdx];
+      if (!state) {
+        this.debug('no transition from ' + this._hwState + '[' + evt + ']');
+        return;
       }
-    },
 
-    /**
-     * Basing on the new value of NFC Setting computes new NFC HW state
-     * and uses {@link NfcManager#_changeHardwareState} to set it
-     * @memberof NfcManager.prototype
-     * @param {boolean} enabled - NFC setting value
-     */
-    _nfcSettingsChanged: function nm_nfcSettingsChanged(enabled) {
-      this._debug('_nfcSettingsChanged, nfc.enabled: ' + enabled);
+      this._hwState = state;
+      this._notifyNfcStateChanged();
+      this._changeNfcHwState();
+    },
 
-      if (this.isActive() === enabled || this.isInTransition()) {
-        this._debug('_nfcSettingsChanged ignoring settings change');
-        return;
+    _notifyNfcStateChanged: function() {
+      switch (this._hwState) {
+        case 'disabling':
+        case 'enabling':
+          this.writeSetting({ 'nfc.status': this._hwState });
+          break;
+        case 'disabled':
+        case 'enabled':
+          this.writeSetting({ 'nfc.status': this._hwState });
+          this.icon && this.icon.update();
+          break;
       }
-
-      var state = !enabled ? this.NFC_HW_STATE.DISABLING :
-        (Service.locked ? this.NFC_HW_STATE.DISABLE_DISCOVERY :
-                          this.NFC_HW_STATE.ENABLING);
-      this._changeHardwareState(state);
     },
 
-    /**
-     * Triggers DOM request to change NFC Hardware state
-     * @memberof NfcManager.prototype
-     * @param {string} state - new hardware state, one of
-     * {@link NfcManager#NFC_HW_STATE}
-     */
-    _changeHardwareState: function nm_changeHardwareState(state) {
-      this._debug('_changeHardwareState - state : ' + state);
-      this._hwState = state;
+    _changeNfcHwState: function() {
       var nfcdom = window.navigator.mozNfc;
       if (!nfcdom) {
+        this._doNfcStateTransition('hw-change-failure');
```

Comment:
We will always have navigator.mozNfc is NfcManager is loaded. So I wonder if we need this.